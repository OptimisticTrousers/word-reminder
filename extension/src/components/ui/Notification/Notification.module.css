.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --bkg-color: rgb(220, 252, 231);
  --base-text-color: rgb(20, 83, 45);
  --btn-text-color: rgb(22, 163, 74);
  --icon-color: rgb(21, 128, 61);
  background-color: var(--bkg-color);
}

.notification--error {
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--base-text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--btn-text-color);
  padding: 0.375rem;
  background-color: var(--bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
