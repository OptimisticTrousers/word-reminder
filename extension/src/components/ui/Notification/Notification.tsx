import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
