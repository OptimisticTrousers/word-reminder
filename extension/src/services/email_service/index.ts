export { emailService } from "./email_service";
