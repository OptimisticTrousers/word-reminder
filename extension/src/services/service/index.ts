export type { AuthParams, Params } from "./service";
export { service } from "./service";
