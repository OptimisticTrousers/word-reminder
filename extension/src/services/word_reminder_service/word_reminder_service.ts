import { AutoWordReminderParams, ManualWordReminderParams } from "common";

import { service, Params } from "../service";

export const wordReminderService = (function (service) {
  const { get, post, put, remove, VITE_API_DOMAIN } = service;

  function getWordReminderList(userId: string, params: Params) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/wordReminders`,
      params,
      options: {
        credentials: "include",
      },
    });
  }

  function getWordReminder(userId: string, wordReminderId: string) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/wordReminders/${wordReminderId}`,
      options: {
        credentials: "include",
      },
    });
  }

  function createWordReminder({
    userId,
    body,
  }: {
    userId: string;
    body: ManualWordReminderParams | AutoWordReminderParams;
  }) {
    return post({
      url: `${VITE_API_DOMAIN}/users/${userId}/wordReminders`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function deleteWordReminders(userId: string) {
    return remove({
      url: `${VITE_API_DOMAIN}/users/${userId}/wordReminders`,
      options: { credentials: "include" },
    });
  }

  function updateWordReminder({
    userId,
    wordReminderId,
    body,
  }: {
    userId: string;
    wordReminderId: string;
    body: Omit<ManualWordReminderParams, "auto">;
  }) {
    return put({
      url: `${VITE_API_DOMAIN}/users/${userId}/wordReminders/${wordReminderId}`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function deleteWordReminder({
    userId,
    wordReminderId,
  }: {
    userId: string;
    wordReminderId: string;
  }) {
    return remove({
      url: `${VITE_API_DOMAIN}/users/${userId}/wordReminders/${wordReminderId}`,
      options: { credentials: "include" },
    });
  }

  return {
    createWordReminder,
    deleteWordReminder,
    deleteWordReminders,
    getWordReminder,
    getWordReminderList,
    updateWordReminder,
  };
})(service);
