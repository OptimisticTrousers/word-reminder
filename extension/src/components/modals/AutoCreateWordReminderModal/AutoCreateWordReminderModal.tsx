import { Order, User } from "common";
import CSSModules from "react-css-modules";
import { useOutletContext } from "react-router-dom";
import { useNotificationError } from "../../../hooks/useNotificationError";
import { ToggleModal } from "../types";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import { wordReminderService } from "../../../services/word_reminder_service";
import { useContext } from "react";
import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import { ModalContainer } from "../ModalContainer";
import styles from "./AutoCreateWordReminderModal.module.css";
import { AddToDate } from "../../ui/AddToDate";

interface Props {
  searchParams: URLSearchParams;
  toggleModal: ToggleModal;
}

export const AutoCreateWordReminderModal = CSSModules(
  function ({ searchParams, toggleModal }: Props) {
    const { user }: { user: User } = useOutletContext();
    const userId = user.id;
    const { showNotification } = useContext(NotificationContext);
    const { showNotificationError } = useNotificationError();
    const queryClient = useQueryClient();
    const { isPending, mutate } = useMutation({
      mutationFn: wordReminderService.createWordReminder,
      onSuccess: () => {
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          AUTO_CREATE_WORD_REMINDER_NOTIFICATION_MSGS.autoCreateWordReminder()
        );
        const searchParamsObject = Object.fromEntries(searchParams);
        queryClient.invalidateQueries({
          queryKey: ["wordReminders", searchParamsObject],
          exact: true,
        });
      },
      onError: showNotificationError,
      onSettled: () => {
        toggleModal();
      },
    });

    function handleCreate(formData: FormData) {
      mutate({
        userId,
        body: {
          auto: true,
          reminder: {
            minutes: Number(formData.get("reminder-minutes") as string),
            hours: Number(formData.get("reminder-hours") as string),
            days: Number(formData.get("reminder-days") as string),
            weeks: Number(formData.get("reminder-weeks") as string),
            months: Number(formData.get("reminder-months") as string),
          },
          create_now: Boolean(formData.get("create_now") as string),
          duration: {
            minutes: Number(formData.get("duration-minutes") as string),
            hours: Number(formData.get("duration-hours") as string),
            days: Number(formData.get("duration-days") as string),
            weeks: Number(formData.get("duration-weeks") as string),
            months: Number(formData.get("duration-months") as string),
          },
          word_count: Number(formData.get("word_count") as string),
          is_active: Boolean(formData.get("is_active") as string),
          has_reminder_onload: Boolean(
            formData.get("has_reminder_onload") as string
          ),
          has_learned_words: Boolean(
            formData.get("has_learned_words") as string
          ),
          order: formData.get("order") as Order,
        },
      });
    }

    return (
      <ModalContainer
        title="Automatically Create Word Reminder"
        toggleModal={toggleModal}
      >
        <form styleName="modal__form" action={handleCreate}>
          <AddToDate legend="Reminder" disabled={false} />
          <AddToDate legend="Duration" disabled={false} />
          <div styleName="modal__control">
            <label styleName="modal__label" htmlFor="word_count">
              Word Count
            </label>
            <input
              styleName="modal__input"
              type="number"
              id="word_count"
              name="word_count"
              required
              max={99}
            />
          </div>
          <fieldset styleName="modal__fieldset">
            <legend styleName="modal__legend">Options</legend>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="create_now">
                Create Now
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="create_now"
                name="create_now"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="is_active">
                Is Active
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="is_active"
                name="is_active"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="has_reminder_onload">
                Has Reminder Onload
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="has_reminder_onload"
                name="has_reminder_onload"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="has_learned_words">
                Has Learned Words
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="has_learned_words"
                name="has_learned_words"
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="order">
                Order
              </label>
              <select
                styleName="modal__select"
                name="order"
                id="order"
                required
                defaultValue={Order.Random}
              >
                <option styleName="modal__option" value={Order.Newest}>
                  Newest
                </option>
                <option styleName="modal__option" value={Order.Oldest}>
                  Oldest
                </option>
                <option styleName="modal__option" value={Order.Random}>
                  Random
                </option>
              </select>
            </div>
          </fieldset>
          <button styleName="modal__button" type="submit" disabled={isPending}>
            Create
          </button>
        </form>
      </ModalContainer>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);

const AUTO_CREATE_WORD_REMINDER_NOTIFICATION_MSGS = {
  autoCreateWordReminder: () => {
    return "Word reminders will now be created automatically!";
  },
};
