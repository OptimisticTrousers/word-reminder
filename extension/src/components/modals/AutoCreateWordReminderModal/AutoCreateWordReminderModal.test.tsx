import { render, screen } from "@testing-library/react";
import { createRoutesStub, Outlet } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { AutoCreateWordReminderModal } from "./AutoCreateWordReminderModal";
import userEvent from "@testing-library/user-event";
import { wordReminderService } from "../../../services/word_reminder_service";
import { Order } from "common";
import { NotificationProvider } from "../../../context/Notification";
import { Mock } from "vitest";

vi.mock("../ModalContainer/ModalContainer");

describe("AutoCreateWordReminderModal component", () => {
  const testUser = {
    id: "1",
  };

  function setup({
    toggleModal,
    queryClient,
  }: {
    toggleModal: Mock;
    queryClient: QueryClient;
  }) {
    const searchParams = new URLSearchParams();
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <NotificationProvider>
                  <QueryClientProvider client={queryClient}>
                    <AutoCreateWordReminderModal
                      searchParams={searchParams}
                      toggleModal={toggleModal}
                    />
                  </QueryClientProvider>
                </NotificationProvider>
              );
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/"]} />),
    };
  }

  const wordReminder = {
    id: "1",
  };

  const status = 200;

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders form with all inputs", async () => {
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const { asFragment } = setup({ toggleModal: mockToggleModal, queryClient });

    const reminder = screen.getByLabelText("Reminder", { selector: "input" });
    const duration = screen.getByLabelText("Duration", { selector: "input" });
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createNow = screen.getByLabelText("Create Now", {
      selector: "input",
    });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const hasLearnedWords = screen.getByLabelText("Has Learned Words", {
      selector: "input",
    });
    const order = screen.getByLabelText("Order", { selector: "select" });
    const notification = screen.queryByRole("dialog");
    const modalHeading = screen.getByTestId("modal-heading");
    expect(modalHeading).toBeInTheDocument();
    expect(modalHeading).toHaveTextContent(
      "Automatically Create Word Reminder"
    );
    expect(notification).not.toBeInTheDocument();
    expect(reminder).toBeInTheDocument();
    expect(duration).toBeInTheDocument();
    expect(wordCount).toBeInTheDocument();
    expect(createNow).toBeInTheDocument();
    expect(createNow).toBeChecked();
    expect(isActive).toBeInTheDocument();
    expect(isActive).toBeChecked();
    expect(hasReminderOnload).toBeInTheDocument();
    expect(hasReminderOnload).toBeChecked();
    expect(hasLearnedWords).toBeInTheDocument();
    expect(hasLearnedWords).not.toBeChecked();
    expect(order).toBeInTheDocument();
    expect(order).toHaveValue(Order.Random);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the functions to automatically create a word reminder", async () => {
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        return { json: { wordReminder }, status };
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const reminder = screen.getByLabelText("Reminder", { selector: "input" });
    const duration = screen.getByLabelText("Duration", { selector: "input" });
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createButton = screen.getByRole("button", { name: "Create" });
    const createNow = screen.getByLabelText("Create Now", {
      selector: "input",
    });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const hasLearnedWords = screen.getByLabelText("Has Learned Words", {
      selector: "input",
    });
    const order = screen.getByLabelText("Order", { selector: "select" });

    await user.type(reminder, "1 hour");
    await user.type(duration, "2 weeks");
    await user.type(wordCount, "7");
    await user.click(createNow);
    await user.click(isActive);
    await user.click(hasReminderOnload);
    await user.click(hasLearnedWords);
    await user.selectOptions(order, Order.Oldest);
    await user.click(createButton);

    const notification = screen.getByRole("dialog", {
      name: "Word reminders will now be created automatically!",
    });
    expect(notification).toBeInTheDocument();
    expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
    expect(mockInvalidateQueries).toHaveBeenCalledWith({
      queryKey: ["wordReminders", Object.fromEntries(new URLSearchParams())],
      exact: true,
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockCreateWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        auto: true,
        reminder: "1 hour",
        create_now: false,
        duration: "2 weeks",
        word_count: 7,
        is_active: false,
        has_reminder_onload: false,
        has_learned_words: true,
        order: Order.Oldest,
      },
    });
  });

  it("calls the functions to show a notification error", async () => {
    const message = "Bad Request.";
    const status = 400;
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const reminder = screen.getByLabelText("Reminder", { selector: "input" });
    const duration = screen.getByLabelText("Duration", { selector: "input" });
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(reminder, "1 hour");
    await user.type(duration, "2 weeks");
    await user.type(wordCount, "7");
    await user.click(createButton);

    const notification = screen.getByRole("dialog", { name: message });
    expect(notification).toBeInTheDocument();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockCreateWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        auto: true,
        reminder: "1 hour",
        duration: "2 weeks",
        create_now: true,
        word_count: 7,
        is_active: true,
        has_reminder_onload: true,
        has_learned_words: false,
        order: Order.Random,
      },
    });
  });

  it("disables the create button when the mutation is loading", async () => {
    const delay = 500;
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json: { wordReminder }, status });
          }, delay);
        });
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const reminder = screen.getByLabelText("Reminder", { selector: "input" });
    const duration = screen.getByLabelText("Duration", { selector: "input" });
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(reminder, "1 hour");
    await user.type(duration, "2 weeks");
    await user.type(wordCount, "7");
    await user.click(createButton);

    const notification = screen.queryByRole("dialog");
    expect(notification).not.toBeInTheDocument();
    expect(createButton).toBeDisabled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
    expect(mockToggleModal).not.toHaveBeenCalled();
    expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockCreateWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        auto: true,
        reminder: "1 hour",
        duration: "2 weeks",
        word_count: 7,
        is_active: true,
        create_now: true,
        has_reminder_onload: true,
        has_learned_words: false,
        order: Order.Random,
      },
    });
  });

  it("calls the functions to close a modal", async () => {
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        return { json: { wordReminder }, status };
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });

    const closeButton = screen.getByRole("button", { name: "Close modal" });
    await user.click(closeButton);

    const notification = screen.queryByRole("dialog");
    expect(notification).not.toBeInTheDocument();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockCreateWordReminder).not.toHaveBeenCalled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
  });

  describe("form validation", () => {
    it("only allows the user to enter up to 10 characters for reminder", async () => {
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder", { selector: "input" });
      const duration = screen.getByLabelText("Duration", { selector: "input" });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminder, "10000 hours");
      await user.type(duration, "2 weeks");
      await user.type(wordCount, "7");
      await user.click(createButton);

      const notification = screen.getByRole("dialog");
      expect(notification).toBeInTheDocument();
      expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
      expect(mockCreateWordReminder).toHaveBeenCalledWith({
        userId: testUser.id,
        body: {
          auto: true,
          reminder: "10000 hour",
          create_now: true,
          duration: "2 weeks",
          word_count: 7,
          is_active: true,
          has_reminder_onload: true,
          has_learned_words: false,
          order: Order.Random,
        },
      });
      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
      expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
      expect(mockInvalidateQueries).toHaveBeenCalledWith({
        queryKey: ["wordReminders", Object.fromEntries(new URLSearchParams())],
        exact: true,
      });
    });

    it("only allows the user to enter up to 99 for wordCount", async () => {
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder", { selector: "input" });
      const duration = screen.getByLabelText("Duration", { selector: "input" });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminder, "10000 hours");
      await user.type(duration, "2 weeks");
      await user.type(wordCount, "100");
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockCreateWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("only allows the user to enter up to 10 characters for duration", async () => {
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder", { selector: "input" });
      const duration = screen.getByLabelText("Duration", { selector: "input" });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminder, "1 hour");
      await user.type(duration, "20000 weeks");
      await user.type(wordCount, "7");
      await user.click(createButton);

      const notification = screen.getByRole("dialog");
      expect(notification).toBeInTheDocument();
      expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
      expect(mockCreateWordReminder).toHaveBeenCalledWith({
        userId: testUser.id,
        body: {
          auto: true,
          reminder: "1 hour",
          create_now: true,
          duration: "20000 week",
          word_count: 7,
          is_active: true,
          has_reminder_onload: true,
          has_learned_words: false,
          order: Order.Random,
        },
      });
      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
      expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
      expect(mockInvalidateQueries).toHaveBeenCalledWith({
        queryKey: ["wordReminders", Object.fromEntries(new URLSearchParams())],
        exact: true,
      });
    });

    it("does not allow the user to submit when wordCount is empty", async () => {
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder", { selector: "input" });
      const duration = screen.getByLabelText("Duration", { selector: "input" });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminder, "1 hour");
      await user.type(duration, "2 weeks");
      await user.clear(wordCount);
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockCreateWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("does not allow the user to submit when reminder is empty", async () => {
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder", { selector: "input" });
      const duration = screen.getByLabelText("Duration", { selector: "input" });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.clear(reminder);
      await user.type(duration, "2 weeks");
      await user.type(wordCount, "7");
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockCreateWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("does not allow the user to submit when duration is empty", async () => {
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder", { selector: "input" });
      const duration = screen.getByLabelText("Duration", { selector: "input" });
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminder, "1 hour");
      await user.clear(duration);
      await user.type(wordCount, "7");
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockCreateWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });
  });
});
