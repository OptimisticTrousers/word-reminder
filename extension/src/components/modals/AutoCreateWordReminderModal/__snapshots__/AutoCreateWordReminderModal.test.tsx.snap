// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AutoCreateWordReminderModal component > renders form with all inputs 1`] = `
<DocumentFragment>
  <div>
    <h2
      data-testid="modal-heading"
    >
      Automatically Create Word Reminder
    </h2>
    <button
      aria-label="Close modal"
    >
      X
    </button>
    <div>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal__form_1a29e7"
      >
        <div
          class="_modal__control_1a29e7"
        >
          <label
            class="_modal__label_1a29e7"
            for="reminder"
          >
            Reminder
          </label>
          <input
            class="_modal__input_1a29e7"
            id="reminder"
            maxlength="10"
            name="reminder"
            placeholder="1 hour (indicating every hour), 30 minutes (indicating every thirty minutes)"
            required=""
            type="text"
          />
        </div>
        <div
          class="_modal__control_1a29e7"
        >
          <label
            class="_modal__label_1a29e7"
            for="duration"
          >
            Duration
          </label>
          <input
            class="_modal__input_1a29e7"
            id="duration"
            maxlength="10"
            name="duration"
            placeholder="1 week, 2 weeks, 10 days"
            required=""
            type="text"
          />
        </div>
        <div
          class="_modal__control_1a29e7"
        >
          <label
            class="_modal__label_1a29e7"
            for="word_count"
          >
            Word Count
          </label>
          <input
            class="_modal__input_1a29e7"
            id="word_count"
            max="99"
            name="word_count"
            required=""
            type="number"
          />
        </div>
        <fieldset
          class="_modal__fieldset_1a29e7"
        >
          <legend
            class="_modal__legend_1a29e7"
          >
            Options
          </legend>
          <div
            class="_modal__control_1a29e7"
          >
            <label
              class="_modal__label_1a29e7"
              for="create_now"
            >
              Create Now
            </label>
            <input
              checked=""
              class="_modal__input_1a29e7"
              id="create_now"
              name="create_now"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_1a29e7"
          >
            <label
              class="_modal__label_1a29e7"
              for="is_active"
            >
              Is Active
            </label>
            <input
              checked=""
              class="_modal__input_1a29e7"
              id="is_active"
              name="is_active"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_1a29e7"
          >
            <label
              class="_modal__label_1a29e7"
              for="has_reminder_onload"
            >
              Has Reminder Onload
            </label>
            <input
              checked=""
              class="_modal__input_1a29e7"
              id="has_reminder_onload"
              name="has_reminder_onload"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_1a29e7"
          >
            <label
              class="_modal__label_1a29e7"
              for="has_learned_words"
            >
              Has Learned Words
            </label>
            <input
              class="_modal__input_1a29e7"
              id="has_learned_words"
              name="has_learned_words"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_1a29e7"
          >
            <label
              class="_modal__label_1a29e7"
              for="order"
            >
              Order
            </label>
            <select
              class="_modal__select_1a29e7"
              id="order"
              name="order"
              required=""
            >
              <option
                class="_modal__option_1a29e7"
                value="newest"
              >
                Newest
              </option>
              <option
                class="_modal__option_1a29e7"
                value="oldest"
              >
                Oldest
              </option>
              <option
                class="_modal__option_1a29e7"
                selected=""
                value="random"
              >
                Random
              </option>
            </select>
          </div>
        </fieldset>
        <button
          class="_modal__button_1a29e7"
          type="submit"
        >
          Create
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
